File: Assets/TopKong/Scripts/Fighters/Locomotion.cs
using UnityEngine;

namespace TopKong
{
    /// <summary>
    /// Движение и разворот управляемого бойца.
    ///
    /// Заменил прежний BalanceController целиком. Тот держал тело на ногах пружинами
    /// и моментами: подвеска под тазом, PD-контроллер вертикали, отдельный режим
    /// подъёма с земли. Всё это было нужно только потому, что тело всё время было
    /// ragdoll'ом. Теперь под управлением тело кинематическое, стоять ему не надо
    /// помогать — и здесь остаётся ровно то, что и должно быть в контроллере
    /// персонажа: скорость по земле, разворот и проверка опоры.
    /// </summary>
    public class Locomotion
    {
        readonly Fighter _f;
        readonly GameTuning _t;
        readonly Arena _arena;

        public bool Grounded { get; private set; }
        public float PlanarSpeed { get; private set; }

        public Locomotion(Fighter fighter, GameTuning tuning, Arena arena)
        {
            _f = fighter;
            _t = tuning;
            _arena = arena;
        }

        public void Tick(float dt, bool controlEnabled)
        {
            var body = _f.RootBody;
            if (body == null) return;

            ProbeGround(body);
            ApplyMovement(body, dt, controlEnabled);
            ApplyFacing(body, dt);
        }

        void ProbeGround(Rigidbody body)
        {
            Grounded = false;
            if (_arena.Ground == null) return;

            // Луч проверяется только против коллайдера арены: так не нужны ни слои,
            // ни фильтрация собственных коллайдеров бойца.
            var ray = new Ray(body.position + Vector3.up * 0.4f, Vector3.down);
            Grounded = _arena.Ground.Raycast(ray, out _, 0.6f);
        }

        void ApplyMovement(Rigidbody body, float dt, bool controlEnabled)
        {
            Vector2 move = controlEnabled ? _f.MoveInput : Vector2.zero;
            Vector3 wish = new Vector3(move.x, 0f, move.y);
            if (wish.sqrMagnitude > 1f) wish.Normalize();

            Vector3 velocity = RB.Vel(body);
            Vector3 planar = new Vector3(velocity.x, 0f, velocity.z);

            if (Grounded)
            {
                Vector3 target = wish * _t.maxRunSpeed;
                float rate = wish.sqrMagnitude > 0.01f ? _t.moveAccel : _t.moveBrake;
                planar = Vector3.MoveTowards(planar, target, rate * dt);
            }
            else if (wish.sqrMagnitude > 0.01f)
            {
                // В воздухе управление слабое: сбитый боец должен долетать до края,
                // а не выруливать обратно на арену.
                planar = Vector3.MoveTowards(planar, wish * _t.maxRunSpeed, _t.airControl * dt);
            }

            PlanarSpeed = planar.magnitude;
            RB.SetVel(body, new Vector3(planar.x, velocity.y, planar.z));
        }

        /// <summary>
        /// Разворот к прицелу. Тело кинематическое, поэтому поворот задаётся напрямую
        /// поворотом корня — никакой инерции дубины, за которую цеплялась прошлая схема.
        /// </summary>
        void ApplyFacing(Rigidbody body, float dt)
        {
            Vector3 want = _f.FacingTarget;
            want.y = 0f;
            if (want.sqrMagnitude < 0.0001f) return;

            Quaternion target = Quaternion.LookRotation(want.normalized, Vector3.up);
            // Поворот пишется прямо в трансформ, а не через MoveRotation: у корня стоит
            // FreezeRotation, и физический поворот он бы попросту не пропустил.
            _f.transform.rotation = Quaternion.RotateTowards(
                _f.transform.rotation, target, _t.turnSpeed * dt);
        }
    }
}

File: docs/src/locomotion.js
import { tuning as T } from 'tk/tuning.js';
import { DEG, RAD, deltaAngle, moveTowards, lerp } from 'tk/mathx.js';
import { P } from 'tk/body.js';

// Движение и разворот. Теперь это намерение, а не результат.
//
// Раньше здесь двигался сам трансформ бойца, и куда его поставили — там он
// и оказывался. Теперь тело физическое: локомоция задаёт ядру желаемую
// скорость и крутит целевой разворот, а доедет ли туда тело и как оно при
// этом завалится — решают мышцы и связи.
//
// Корень при этом остаётся авторитетным: ввод меняет намерение мгновенно.
// Именно это удерживает управление от вязкости, из-за которой мы ушли
// от физического тела в Unity-версии, — отстаёт картинка, а не реакция.

export class Locomotion {
  constructor(fighter, arena) {
    this.f = fighter;
    this.arena = arena;
    this.grounded = false;
    this.planarSpeed = 0;
    /** Заказанная вводом скорость в мире. Её читает походка. */
    this.wantX = 0;
    this.wantZ = 0;
    /** Градусов в секунду. Хранится отдельно, чтобы у разворота был разгон. */
    this.yawSpeed = 0;
  }

  tick(dt, controlEnabled) {
    this.probeGround();
    this.measureSpeed(dt);
    this.applyMovement(dt, controlEnabled);
    this.applyFacing(dt, controlEnabled);
  }

  /**
   * Опора есть, если хотя бы одна стопа стоит на настиле. Проверять по тазу,
   * как раньше, больше нельзя: таз живёт своей жизнью и на шаге подпрыгивает.
   */
  probeGround() {
    const p = this.f.body.pos;
    this.grounded = this.footDown(p[P.FootL]) || this.footDown(p[P.FootR]);
  }

  footDown(foot) {
    return foot.y < 0.22 && this.arena.isOverDeck(foot.x, foot.z, -0.05);
  }

  /** Скорость меряется по тазу: это честная скорость тела, а не заказанная. */
  measureSpeed(dt) {
    const p = this.f.body.pos[P.Hips];
    const q = this.f.body.prev[P.Hips];
    this.planarSpeed = Math.hypot(p.x - q.x, p.z - q.z) / Math.max(1e-5, dt);
  }

  applyMovement(dt, controlEnabled) {
    const f = this.f;
    let wx = 0;
    let wz = 0;
    if (controlEnabled) {
      wx = f.moveInput.x;
      wz = f.moveInput.y;
      const len = Math.hypot(wx, wz);
      if (len > 1) { wx /= len; wz /= len; }
    }

    // Пронос забирает управление: удар должен чего-то стоить.
    let scale = 1;
    if (f.swing.striking) scale = T.swingMoveLock;
    else if (f.swing.held) scale = T.chargeMoveSlow;

    // Вбок и назад боец идёт медленнее, и это не условность жанра, а прямое
    // следствие устройства ног. Приставным шагом стопа уходит вбок недалеко:
    // наружу мешает досягаемость, внутрь — вторая нога. Замерено, что на
    // полной скорости вбок обе ноги хотят оторваться разом, одна ждёт
    // очереди, и её опору успевает утащить на 97 см — вдвое дальше,
    // чем нога вообще достаёт.
    const moving = wx !== 0 || wz !== 0;
    let dirScale = 1;
    if (moving) {
      const sin = Math.sin(f.yaw);
      const cos = Math.cos(f.yaw);
      const along = (wx * sin + wz * cos) / Math.max(1e-6, Math.hypot(wx, wz));
      dirScale = along >= 0
        ? lerp(T.strafeSpeed, 1, along)
        : lerp(T.strafeSpeed, T.backSpeed, -along);
    }

    const speed = T.maxRunSpeed * scale * dirScale;
    // Заказанная скорость нужна походке отдельно от настоящей. Пока обе
    // стопы на настиле, цельные ноги держат таз намертво: настоящая
    // скорость там ноль, и по ней шаг не начнётся никогда.
    this.wantX = wx * speed;
    this.wantZ = wz * speed;
    // В воздухе управление слабое: сбитый должен долетать до края,
    // а не выруливать обратно на арену.
    const rate = this.grounded
      ? (moving ? T.moveAccel : T.moveBrake)
      : T.airControl;

    f.body.drive(wx * speed, wz * speed, rate, dt);
  }

  /**
   * Разворот к прицелу — с разгоном и торможением.
   *
   * Плоская скорость в 420 градусов в секунду читалась как мгновенный
   * доворот детали. Здесь у разворота есть угловое ускорение, а перед целью
   * он тормозит ровно настолько, чтобы успеть остановиться: без этого
   * инерция даёт вечное рыскание вокруг прицела.
   */
  applyFacing(dt, controlEnabled) {
    const f = this.f;
    const want = f.facingTarget;
    const hasTarget = controlEnabled && (want.x * want.x + want.z * want.z > 1e-8);

    let desired = 0;
    if (hasTarget) {
      const target = Math.atan2(want.x, want.z) * RAD;
      const error = deltaAngle(f.yaw * RAD, target);
      // Скорость, с которой ещё успеваем остановиться к цели.
      const brakeSpeed = Math.sqrt(2 * T.turnAccel * Math.abs(error));
      desired = Math.sign(error) * Math.min(T.turnSpeed, brakeSpeed);
    }

    this.yawSpeed = moveTowards(this.yawSpeed, desired, T.turnAccel * dt);
    f.yaw += this.yawSpeed * dt * DEG;
  }

  reset() {
    this.yawSpeed = 0;
    this.planarSpeed = 0;
    this.wantX = 0;
    this.wantZ = 0;
    this.grounded = false;
  }
}
